### PR TITLE
Moved Connection Parsing and Checking to Connection Helper

### PR DESCRIPTION
**Pull request by azoitl**
In order to make the ConnectionBuilder better maintainable and allow it to be used for creating mapped to unmapped FB connections all connection parsing, creation, and checking code is moved to the ConnectionBuilder.

### CODE REVIEW

**Review by oberlehner — APPROVED** [review]

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 1]
```diff
@@ -10,39 +10,71 @@
...
 			this.connectionState = EnumSet.of(ConnectionState.VALID);
-			this.source = source;
-			this.destination = destination;
+			this.sourceString = sourceString;
+			this.destinationString = destinationString;
+			this.connection = connection;
+			this.importer = importer;
+
+			connection.setSource(getConnectionEndPoint(sourceString, false));
+			connection.setDestination(getConnectionEndPoint(destinationString, true));
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java:72]: How about we always resolve and validate connections at the end, also for regular networks? That should save us the extra repair step and also make regular networks immune to the order of XML elements?

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 2]
```diff
@@ -79,21 +111,67 @@ public void validate() {
 				connectionState.remove(ConnectionState.VALID);
 			}
 
-			if (destination != null && destInterfaceList != null) {
+			if (destinationString != null && destInterfaceList != null) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: Why is this condition different to the block for the source above?

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 3]
```diff
@@ -179,82 +253,224 @@ public boolean isEmptyConnection() {
...
 
-		public InterfaceList getSrcInterfaceList() {
+		private InterfaceList getSrcInterfaceList() {
 			return srcInterfaceList;
 		}
 
-		public void setSrcInterfaceList(final InterfaceList srcInterfaceList) {
+		private void setSrcInterfaceList(final InterfaceList srcInterfaceList) {
 			this.srcInterfaceList = srcInterfaceList;
 		}
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: Should we inline these simple field accesses?

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 4]
```diff
@@ -179,82 +253,224 @@ public boolean isEmptyConnection() {
...
+
+			// we need a special treatment for FB's that lost their type
+			if (oppositeEndpoint == null) {
+				getConnectionState().add(ConnectionState.MISSING_TYPE);
+				return null;
+			}
+
+			DataType type = oppositeEndpoint.getType();
+			if (type == null) {
+				type = IecTypes.GenericTypes.ANY;
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: Should this not also use `determineConnectionType(...)` above to properly handle event or adapter connections?

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java (hunk 1 above)
Comment (azoitl) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java:72]: with the work I did here we could reconsider later. But as the order in the xml is defined by the dtd I see less need for all networks.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java (hunk 2 above)
Comment (azoitl) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: I guess that is a copy paste error from the original code. The code here is only migrated from the original FBNetwork importer will fix that.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 5]
```diff
@@ -79,21 +111,67 @@ public void validate() {
 				connectionState.remove(ConnectionState.VALID);
 			}
 
-			if (destination != null && destInterfaceList != null) {
+			if (destinationEndpoint != null && destInterfaceList != null) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: Should this not rather just be:
```suggestion
			if (destInterfaceList != null) {
```
as above: https://github.com/eclipse-4diac/4diac-ide/blob/cd46a5449024fe2847367cd23856d8ba7f783ee7/plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java#L100-L101
or am I missing something?

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java (hunk 5 above)
Comment (oberlehner) [reply plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]: I think martin is right

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java [hunk 6]
```diff
@@ -79,21 +111,67 @@ public void validate() {
 				connectionState.remove(ConnectionState.VALID);
 			}
 
-			if (destination != null && destInterfaceList != null) {
+			if (destinationEndpoint != null) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/dataimport/ConnectionHelper.java]:
```suggestion
			if (destInterfaceList != null) {
```

**Review by mx990 — CHANGES_REQUESTED** [review]

**Review by mx990 — APPROVED** [review]

**Pull request merged**